### PR TITLE
Develop doc

### PR DESCRIPTION
**Pull request by mexanick**
Add some documentation, not yet complete, but sufficient to serve as a template/example
## Description

Few modules documented, automated tools for documentation has been included in CI routine. 
### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Motivation and Context

We need proper docs
## How Has This Been Tested?

Tested locally + travis-docker CI routine
### Screenshots (if appropriate):
screen shot 2018-05-29 at 11 35 12 am
screen shot 2018-05-29 at 11 34 52 am
screen shot 2018-05-29 at 11 35 58 am

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

**Review by jsturdy — CHANGES_REQUESTED** [review]
* Reason for `reg_utils/doc` rather than `reg_utils/python/<package>/doc`?
  * Only pertinent if `docs` ship with either the `RPM` (as `man` pages), or the `pip` installable package...
* What is the purpose of the multiple `.gitignore` files in the `doc` tree structure?
  * A top-level `.gitignore` will handle everything in the subdirectories and a cursory glance didn't reveal any differences
  * If there's a need for a `doc` specific `.gitignore`, it can be placed in the `doc` folder, and there's no further need for a `.gitignore` at each nested level
* Better to have the `Makefile` create the necessary (currently empty) directories (e.g., `_build`, `_templates`, `_static`) for the docs, and not have them part of the repository

**Comment by mexanick** [comment]
> Reason for reg_utils/doc rather than reg_utils/python/<package>/doc?

Few reasons: 

- we will have also the c++ part (going to start migration off of this PR) for which I'd like to integrate the documentation as well
- we have an overhead with introduction which is for both packages (yet has to be properly filled as well as the `reg_generator` doc)
- actually, the docstrings themselves are in python source files, it is only general story-like documentation plus some config is upstream
- and finally `reg_interface` and `reg_generator` are subpackaged and latter will dependent on the former

>What is the purpose of the multiple .gitignore files in the doc tree structure?
>Better to have the Makefile create the necessary (currently empty) directories (e.g., _build, _templates, _static) for the docs, and not have them part of the repository

This two questions are connected. I didn't modify the default `Makefile` created by `sphinx`, so that's why I added .gitignores (required to keep folder structure in the repository). I'll look into this. Let me know if you have a ready solution

File: doc/Makefile [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+	@echo "  text       to make text files"
+	@echo "  man        to make manual pages"
+	@echo "  texinfo    to make Texinfo files"
+	@echo "  info       to make Texinfo files and run them through makeinfo"
+	@echo "  gettext    to make PO message catalogs"
+	@echo "  changes    to make an overview of all changed/added/deprecated items"
+	@echo "  linkcheck  to check all external links for integrity"
+	@echo "  doctest    to run all doctests embedded in the documentation (if enabled)"
+
+clean:
```
Comment (jsturdy) [inline doc/Makefile]: You can do the following:

* Add `setup` as a `.PHONY` target above
* Change:
```Makefile
setup:
	mkdir -p $(BUILDDIR)
# If content in _templates and _static is not generated and needs to be version controlled, then comment the following
	mkdir -p _templates
	mkdir -p _static
clean:
	-rm -rf $(BUILDDIR)
# If content in _templates and _static is not generated and needs to be version controlled, then comment the following
	-rm -rf _templates
	-rm -rf _static
```
* Add `setup` as a dependency of all `doc` making targets

**Review by jsturdy — CHANGES_REQUESTED** [review]
OK, I see.  I'd still be in favour of putting the `python` package docs inside the respective package, the `c/c++` docs a the top level, and worrying about the deployment in a system wide `Makefile` (so that all the docs from the `reg_utils` repository are deployed to the same `readthedocs` site) but this can probably be sorted out later.  The main reason for this is per package docs as mentioned.

File: doc/Makefile (hunk 1 above)
Comment (mexanick) [reply doc/Makefile]: I decided to leave the sphinx makefiles as the are. At the moment we don't need anything from _static or _template and it is only a warning, but if we will use something there at some point in the future, we will need to add it to github as those files will have versions and could not be generated

**Comment by mexanick** [comment]
I thought a new push will restart the CI build... However I found it cancelled and had to restart it manually. Is it expected? Is it possible to change this kind of behavior to automatically restart the build check when new commits arrive?

**Comment by jsturdy** [comment]
The current setting is to trigger a build on all `push` and `pr` events; a more recent `push` will cancel the previous build, so your new `push` *should* have triggered a new build, but there can be problems if too many events come in quick succession, i.e., a `tag` and `push` event arrive because you did:
```
git add <changed files>
git ci -m "something changed"
git tag -a -m "note about new tag" vx.y.z
git push origin # pushes commits and tags
```
In cases like this, I've seen the `branch` build triggered, but not the `tag` build...

**Review by jsturdy — CHANGES_REQUESTED** [review]
Another quick couple of things
* Still see the `.gitignore`s inside the `doc/` folders, not used as `git` placeholders here, as those folders aren't empty
* The `sphinx` `Makefile` is generic, no?  Should we add it to the `gembuild` repo and copy it into `doc` via the package `Makefile` so that it is easily reusable from the rest of the projects?

**Comment by mexanick** [comment]
gitignores are removed. 
Yes, the sphinx makefile is generic and your suggestion can be implemented. Could you advise how to do that?

**Comment by jsturdy** [comment]
You can try usual `git` commands:
`git mv path/to/sphinx/Makefile config/mfSphinx.mk`

Then for the inclusion I'll have to take a closer look
Could try in the package `Makefile`, add a line

```
include $(BUILD_HOME)/$(Project)/config//mfSphinx.mk
```
And then let the package `Makefile` drive the whole process (no `Makefile` in the `doc` subdir)

~~I think this would be best, but I'm not sure if there's not some small roadblock I've missed~~
Missed that it's looking for `conf.py` in the directory from where `sphinx-doc` is called.
Adding:
```Makefile
# You can set these variables from the command line.                                                                                                                                                                                                                              
DOCBASE     = $(PackagePath)/doc
SPHINXOPTS  =
SPHINXBUILD = sphinx-build
PAPER       =
SOURCEDIR   = $(DOCBASE)/
BUILDDIR    = $(DOCBASE)/_build

# Internal variables.                                                                                                                                                                                                                                                             
PAPEROPT_a4     = -D latex_paper_size=a4
PAPEROPT_letter = -D latex_paper_size=letter
ALLSPHINXOPTS   = -d $(BUILDDIR)/doctrees $(PAPEROPT_$(PAPER)) $(SPHINXOPTS) $(SOURCEDIR)
# the i18n builder cannot share the environment and doctrees with the others                                                                                                                                                                                                      
I18NSPHINXOPTS  = $(PAPEROPT_$(PAPER)) $(SPHINXOPTS) $(SOURCEDIR)
```
and changing the `clean` target to `cleandoc` in the new `mfSphinx.mk` seems to do the trick, but I didn't test in a `virtualenv` with the package installed, so there's probably one or two other nuances that need sorting out.

**Comment by mexanick** [comment]
The problem with running sphinx build correctly is fixed, however autodoc still fails to import because of missing libraries. I propose to merge this PR and fix doc building in the next PR, where the libraries will be added.

**Review by jsturdy — APPROVED** [review]

**Pull request merged**